### ISSUE TITLE
SQL generated by DSL.dateAdd() is invalid for certain dates in Oracle

### ISSUE THREAD

**Issue by gschlager** [jOOQ/jOOQ#4160]
The generated SQL for `DSL.dateAdd()` is invalid for certain dates in Oracle.

The Java snippet
```java
DSL.dateAdd(DSL.currentDate(), -1, DatePart.MONTH)
```
renders as
```sql
TRUNC(SYSDATE) + NUMTOYMINTERVAL(-1, 'month')
```
Unfortunately, this doesn't work if the calculated date doesn't exist.

The following queries all result in an error:
```sql
-- ORA-01847: day of month must be between 1 and last day of month
SELECT TIMESTAMP '2010-02-29 01:00:00.0' + NUMTOYMINTERVAL(1, 'year') FROM DUAL;

-- ORA-01839: date not valid for month specified
SELECT TIMESTAMP '2015-01-31 01:00:00.0' + NUMTOYMINTERVAL(1, 'month') FROM DUAL;
SELECT TIMESTAMP '2015-03-30 01:00:00.0' + NUMTOYMINTERVAL(-1, 'month') FROM DUAL;
SELECT TIMESTAMP '2015-03-31 01:00:00.0' + NUMTOYMINTERVAL(1, 'month') FROM DUAL;
```

**Comment by gschlager** [comment]
I'm using the following workaround until this is fixed:
```java
DSL.function("ADD_MONTHS", Date.class, DSL.currentDate(), DSL.inline(-1))
```

**Comment by lukaseder** [comment]
Interesting, thank you very much for reporting! We'll look into this asap.

**Comment by lukaseder** [comment]
Hmm, not sure if this needs to be fixed:

```sql
-- ORA-01847: day of month must be between 1 and last day of month
SELECT TIMESTAMP '2010-02-29 01:00:00.0' + NUMTOYMINTERVAL(1, 'year') FROM DUAL;
```

2010 wasn't a leap year. However, when you use 2012, you'll also get ORA-01839 on the `NUMTOYMINTERVAL` addition...

**Comment by lukaseder** [comment]
This is now fixed. The fix will be merged to 3.5.4 (https://github.com/jOOQ/jOOQ/issues/4165), which will be published today. Thanks again for reporting this!
